src/main/core/scheduler/scheduler_policy_host_steal.c

```diff
@@ -374,6 +379,11 @@ static Event* _schedulerpolicyhoststeal_pop(SchedulerPolicy* policy, SimulationT
         g_rw_lock_reader_lock(&data->lock);
         HostStealThreadData* stolenTdata = g_array_index(data->threadList, HostStealThreadData*, stolenTnumber);
         g_rw_lock_reader_unlock(&data->lock);
+
+        /* make sure the workload has been updated for this round */
+        while (!__atomic_load_n(&stolenTdata->isStealable, __ATOMIC_ACQUIRE)) {
```

Comment:
I *think* this looks correct, though could potentially spin more than we'd like, and in general atomic operations outside of implementing higher-level sync primitives is a bit obscure. Maybe a barrier could work instead of a raw atomic?

Reply:
(Or if spinning actually is what you want here, maybe a spin lock?)

Reply:
I don't think a `pthread_barrier` will work here, because I want the bool functionality - I only want the threads waiting for it to turn true to wait, but not the thread that sets it to true. I think `pthread_barrier` requires all threads that decrement the count to wait until it reaches 0.

I was planning to use a `CountDownLatch` but then got excited when I realized I could do it with an atomic with hopefully less synchronization overhead. (The countdown latch uses both a mutex and a condition variable.) I'm using the atomic to prevent a race condition so I don't expect the wait will be very long, so I think spinning is OK.

`pthread_spin_lock` could work, but it's a bit kludgy. If we are using N workers then it's possible that N-1 of them reach this point. I just want the N-1 workers to check/wait for the boolean to become true. If I use a spin lock, I would need to use the unlocked state to represent true, meaning that the N-1 workers would acquire and immediately release the lock every time they want to check if they can steal from another worker. For N workers, this happens N^2 times per round. The atomic seems simpler.

Thoughts?

Reply:
> I don't think a pthread_barrier will work here
...
> I was planning to use a CountDownLatch

Oops, yeah CountDownLatch was what I meant.

> but then got excited when I realized I could do it with an atomic with hopefully less synchronization overhead. (The countdown latch uses both a mutex and a condition variable.) I'm using the atomic to prevent a race condition so I don't expect the wait will be very long, so I think spinning is OK.

I think the fast path without contention would be the ~same? But yeah, the main practical benefit would be avoiding a potential infinite spin. (re?)reading your comment, it sounds like spinning is probably ok.

> pthread_spin_lock could work, but it's a bit kludgy

Hmm, I see what you mean.

Maybe a semaphore (or even Ryan's spinning semaphore) would be a more appropriate abstraction?

Anway, I'm fine with leaving as-is. It makes the code a little more opaque to directly use these low level atomic functions here, but it's not a big deal.

Reply:
OK, I'm leaving as-is (Ryan's spinning semaphore is a great idea but only available in dev branch atm).